wrapper/src/main/java/software/amazon/jdbc/plugin/failover2/FailoverConnectionPlugin.java

```diff
@@ -0,0 +1,619 @@
+    }
+
+    if (isInTransaction || this.pluginService.isInTransaction()) {
+      if (this.pluginManagerService != null) {
+        this.pluginManagerService.setInTransaction(false);
+      }
+      // "Transaction resolution unknown. Please re-configure session state if required and try
+      // restarting transaction."
+      final String errorMessage = Messages.get("Failover.transactionResolutionUnknownError");
+      LOGGER.info(errorMessage);
+      throw new TransactionStateUnknownSQLException();
+    } else {
+      // "The active SQL connection has changed due to a connection failure. Please re-configure
+      // session state if required. "
+      LOGGER.severe(() -> Messages.get("Failover.connectionChangedError"));
+      throw new FailoverSuccessSQLException();
+    }
+  }
+
+  protected void failoverReader() throws SQLException {
+    TelemetryFactory telemetryFactory = this.pluginService.getTelemetryFactory();
+    TelemetryContext telemetryContext = telemetryFactory.openTelemetryContext(
+        TELEMETRY_READER_FAILOVER, TelemetryTraceLevel.NESTED);
+    this.failoverReaderTriggeredCounter.inc();
+
+    this.failoverStartTimeNano = System.nanoTime();
+
+    try {
+      LOGGER.fine(() -> Messages.get("Failover.startReaderFailover"));
+
+      // It's expected that this method synchronously returns when topology is stabilized,
+      // i.e. when cluster control plane has already chosen a new writer.
+      if (!this.pluginService.forceRefreshHostList(false, 0)) {
+        // "Unable to establish SQL connection to reader instance"
+        this.failoverReaderFailedCounter.inc();
+        LOGGER.severe(Messages.get("Failover.unableToConnectToReader"));
+        throw new FailoverFailedSQLException(Messages.get("Failover.unableToConnectToReader"));
+      }
+
+      final List<HostSpec> hosts = this.pluginService.getHosts();
+      Connection readerCandidateConn = null;
+      HostSpec readerCandidate = null;
+      final Set<HostSpec> remainingHosts = new HashSet<>(hosts);
+
+      while (!remainingHosts.isEmpty() && readerCandidateConn == null) {
+        try {
+          readerCandidate =
+              this.pluginService.getHostSpecByStrategy(
+                  remainingHosts.stream().collect(Collectors.toList()),
+                  HostRole.READER,
+                  this.failoverReaderHostSelectorStrategySetting);
+        } catch (UnsupportedOperationException | SQLException ex) {
+          // can't use selected strategy to get a reader host
+          LOGGER.finest("Error: " + ex.getMessage());
+          break;
+        }
+
+        if (readerCandidate == null) {
+          break;
+        }
+
+        try {
+          readerCandidateConn = this.pluginService.connect(readerCandidate, this.properties);
+          if (this.pluginService.getHostRole(readerCandidateConn) != HostRole.READER) {
+            readerCandidateConn.close();
+            readerCandidateConn = null;
+            readerCandidate = null;
```

Comment:
shouldn't we remove this host from the remaining hosts if it isn't valid? Otherwise we might just end up trying to connect to it again
```suggestion
            remainingHosts.remove(readerCandidate);
            readerCandidate = null;
```